### DISCUSSION TITLE
Support for 'Not Ready' status report in Security Systems

### DISCUSSION THREAD

**Discussion by Toby-Mills** [andrei-tatar/node-red-contrib-smartnora#310] [Q&A]
@andrei-tatar I'm busy implementing a flow for my security system. The arming & disarming is working well, but from what I can see the status reporting is limited to 'Disarmed' & 'Armed' (plus configured Arm Levels).
I would like to be able to report on a 'Not Ready' status which should block arming, with specific details of which sensors are not ready. The Google API appears to support this in the StatusReport Trait, but I don't see anything in SmartNORA. Can you confirm if this is supported yet?

**Comment by andrei-tatar** [comment]
@Toby-Mills I will check and get back to you.

**Reply by andrei-tatar** [reply]
@Toby-Mills I think the only way to do this is to use async command execution and report back an error code. This way the actual arm command will be handled by your node-red flow and can either report with the new arm level or with an error code.

Let me get back to you with an example flow.

**Reply by andrei-tatar** [reply]
This below flow should work. Maybe also read the Async command execution doc (https://github.com/andrei-tatar/node-red-contrib-smartnora/blob/master/doc/nodes/async.md)

```json
[{"id":"0cc0c90bccc37286","type":"noraf-security","z":"86c7bfe22f048e9c","devicename":"Security System","roomhint":"","name":"","passthru":false,"nora":"f1e538c1.0315f8","topic":"","twofactor":"off","twofactorpin":"","filter":false,"language":"en","armLevels":[],"asyncCmd":true,"outputs":2,"errorifstateunchaged":false,"x":980,"y":320,"wires":[[],["cd74d4e2750b2bc1"]]},{"id":"a51494baece15d35","type":"inject","z":"86c7bfe22f048e9c","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"isArmed\": false}","payloadType":"json","x":780,"y":340,"wires":[["0cc0c90bccc37286"]]},{"id":"06d99693a13f0d32","type":"inject","z":"86c7bfe22f048e9c","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"isArmed\": true}","payloadType":"json","x":780,"y":300,"wires":[["0cc0c90bccc37286"]]},{"id":"13d1c3f76ae1a8cd","type":"noraf-async","z":"86c7bfe22f048e9c","name":"response","x":1300,"y":320,"wires":[]},{"id":"cd74d4e2750b2bc1","type":"function","z":"86c7bfe22f048e9c","name":"handle","func":"if (msg.payload.command !== 'ArmDisarm') {\n    // should never happen, it's the only command supported by security system\n    return;\n}\n\nif (msg.payload.arm) {\n    //TODO: do your actual check\n    const canArmDevice = Math.random() > 0.5;\n\n    if (!canArmDevice) {\n        // check the list here for better suiting error codes:\n        // https://developers.home.google.com/cloud-to-cloud/intents/errors-exceptions\n        msg.payload = {\n            errorCode: 'securityRestriction',\n        };\n        return;\n    }\n} \n\nconst newState = {\n    isArmed: msg.payload.arm,\n};\n\nif ('armLevel' in msg.payload) {\n    newState.currentArmLevel = msg.payload.armLevel;\n}\n\nmsg.payload = {\n    state: newState,\n};\n\nreturn msg;","outputs":1,"timeout":0,"noerr":0,"initialize":"","finalize":"","libs":[],"x":1150,"y":320,"wires":[["13d1c3f76ae1a8cd"]]},{"id":"f1e538c1.0315f8","type":"noraf-config","name":"NORA","group":"seceani","twofactor":"off","twofactorpin":"","localexecution":true,"structure":"Home","storeStateInContext":true,"disableValidationErrors":false,"sendDeviceNameAndLocation":false}]
```

the async command handler script looks like:

```js
if (msg.payload.command !== 'ArmDisarm') {
    // should never happen, it's the only command supported by security system
    return;
}

if (msg.payload.arm) {
    //TODO: do your actual check
    const canArmDevice = Math.random() > 0.5;

    if (!canArmDevice) {
        // check the list here for better suiting error codes:
        // https://developers.home.google.com/cloud-to-cloud/intents/errors-exceptions
        msg.payload = {
            errorCode: 'securityRestriction',
        };
        return;
    }
} 

const newState = {
    isArmed: msg.payload.arm,
};

if ('armLevel' in msg.payload) {
    newState.currentArmLevel = msg.payload.armLevel;
}

msg.payload = {
    state: newState,
};

return msg;
```

PS: maybe `armFailure` would've been the better error code.

**Reply by Toby-Mills** [reply]
Thanks @andrei-tatar that looks like it will work for reporting a failure when attempting to arm the alarm.
I'm interested in being able to report the status of the alarm before attempting to arm the alarm. It doesn't look like this is currently supported. I'll look into what it could take to support this. Thanks :)

**Reply by andrei-tatar** [reply]
@Toby-Mills as far as I understood, the `StatusReport` trait could help with that, but it seems that it only reports the status when the device is queried. 

https://developers.home.google.com/cloud-to-cloud/traits/statusreport

Anyways, this trait is not yet supported by NORA. If you think it would help you, I can add it when I get a bit of time over the next week or so. But it seems to imply that all the devices you are reporting the status on, need to be added and synced with  google home as well.

**Reply by Toby-Mills** [reply]
Thanks @andrei-tatar if you are able to add support for the trait, that would be very helpful!
